### DISCUSSION TITLE
issue: Unexpected errors with `useForm`'s `resolver` argument with `yup.InferType`

### DISCUSSION THREAD

**Discussion by milotoor** [react-hook-form/react-hook-form#10644] [Q&A] [answered]
### Version Number

7.45.1

### Codesandbox/Expo snack

https://codesandbox.io/p/github/milotoor/react-hook-form-bug-minimal-reproduction/main

### Steps to reproduce

1. Go to the [CodeSandbox](https://codesandbox.io/p/github/milotoor/react-hook-form-bug-minimal-reproduction/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljpbbq51000b3b6pzqua52jn%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljpbbq51000d3b6pmreqpdbn%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljpbbq51000b3b6pzqua52jn%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljpbbq51000a3b6pwmcynufx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522cljpbbq51000b3b6pzqua52jn%2522%252C%2522activeTabId%2522%253A%2522cljpbbq51000a3b6pwmcynufx%2522%257D%252C%2522cljpbbq51000d3b6pmreqpdbn%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljpbbq51000c3b6p1msivjoa%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522build%2522%257D%255D%252C%2522id%2522%253A%2522cljpbbq51000d3b6pmreqpdbn%2522%252C%2522activeTabId%2522%253A%2522cljpbbq51000c3b6p1msivjoa%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
2. After the setup completes, run `npm run build`
3. See error
4. See README file for detailed description of the issue and expected behavior

Alternatively, visit the minimal repro repository [here](https://github.com/milotoor/react-hook-form-bug-minimal-reproduction).

### Expected behaviour

The code should compile. See a detailed discussion in the README, which is copy pasted below in the hidden section.

<details>
<summary>README contents</summary>
This repository provides a minimal reproduction of an issue I've been encountering with the latest versions of `yup`, `react-hook-form` and `@hookform/resolvers`. The repository was bootstrapped with Vite, please forgive any superfluity.

I am looking to create a form component which can be reused in two different contexts: when creating a new object and when editing an existing object. In both cases, the form should be validated against a schema which requires all fields to be provided at submit time. In the former case, most (if not all) form fields will be empty; in the latter case, most should be pre-populated with the existing object values. I am using `yup` to define the schemas and `react-hook-form` to handle the form state and validation. I am using `@hookform/resolvers` to generate a resolver function from the schema which can be passed to `useForm` for validation.

The issue can be seen in `src/App.tsx`. Within that file I define three basic schemas:

```typescript
const schemaRequiredOnly = yup.object({ required: yup.number().required() });
const schemaOptionalOnly = yup.object({ optional: yup.number().optional() });
const schemaBoth = yup.object({
  required: yup.number().required(),
  optional: yup.number().optional(),
});
```

I then define three React components, each of which uses one of the schemas:

```typescript
type ComponentProps<Schema extends yup.ISchema<any>> = {
  defaultValues?: yup.InferType<Schema>;
};

const SchemaRequiredOnly = ({
  defaultValues,
}: ComponentProps<typeof schemaRequiredOnly>) => {
  useForm({
    defaultValues,
    resolver: yupResolver(schemaRequiredOnly),
  });

  return null;
};

const SchemaOptionalOnly = ({
  defaultValues,
}: ComponentProps<typeof schemaOptionalOnly>) => {
  useForm({
    defaultValues,
    resolver: yupResolver(schemaOptionalOnly), // <-- error
  });

  return null;
};

const SchemaBoth = ({ defaultValues }: ComponentProps<typeof schemaBoth>) => {
  useForm({
    defaultValues,
    resolver: yupResolver(schemaBoth), // <-- error
  });

  return null;
};
```

The first component, `SchemaRequiredOnly`, works as expected. The second and third components, however, throw the following errors:

```typescript
src/App.tsx:35:5 - error TS2322: Type 'Resolver<{ optional: number | undefined; }>' is not assignable to type 'Resolver<{ optional?: number | undefined; }, any>'.
  Types of parameters 'values' and 'values' are incompatible.
    Type '{ optional?: number | undefined; }' is not assignable to type '{ optional: number | undefined; }'.

35     resolver: yupResolver(schemaOptionalOnly), // <-- error
```

```typescript
src/App.tsx:44:5 - error TS2322: Type 'Resolver<{ required: number; optional: number | undefined; }>' is not assignable to type 'Resolver<{ optional?: number | undefined; required: number; }, any>'.
  Types of parameters 'values' and 'values' are incompatible.
    Type '{ optional?: number | undefined; required: number; }' is not assignable to type '{ required: number; optional: number | undefined; }'.

44     resolver: yupResolver(schemaBoth), // <-- error
```

The issue appears to be that the `optional` field of the `defaultValues` object is inferred to be optional (i.e. annotated with a `?` in the type generated by `yup.InferType`):

```typescript
type SchemaOptionalOnly = yup.InferType<typeof schemaOptionalOnly>;
// --> { optional?: number | undefined; }
```

By contrast, the resolver object returned from `yupResolver` expects the `optional` field to be provided, though its value may be `undefined`:

```typescript
type OptionalOnlyResolver = typeof optionalOnlyResolver;
// --> Resolver<{ optional: number | undefined; }>
//   --> (values: {
//           optional: number | undefined;
//       }, context: any, options: ResolverOptions<{
//           optional: number | undefined;
//       }>) => ResolverResult<{ optional: number | undefined; }>
```

In other words, the resolver function expects to be called with an object that has an `optional` field (which may be `undefined`) but `useForm` understands that the `optional` field may be omitted entirely.

There are a couple things I can do to remedy this issue:

1. Don't provide `defaultValues` to `useForm` (i.e. `useForm({ resolver: yupResolver(schemaOptionalOnly) })`). This works, but it's not a viable solution for my use case.
2. Ensure that `optional` is included in the `defaultValues`:

  ```typescript
  useForm({
    defaultValues: { ...defaultValues, optional: defaultValues?.optional },
    resolver: yupResolver(schemaOptionalOnly), // <-- no error!
  });
  ```
  
  This is viable but pretty ugly, and obviously scales very poorly.

3. Don't use `InferType`? This is also viable, but it's a bit of a pain to have to manually define the type of the schema when it's readily inferable.

Ultimately, I'm not sure whether this is an issue with `yup`, `react-hook-form`, `@hookform/resolvers` or some combination of the three. The code seems like it should compile, but perhaps there is a better way to accomplish what I'm trying to do. I'm happy to provide any additional information that might be helpful in resolving this issue.

You can run the code with `npm run build` and see the errors in the console. Thanks in advance for any help you can provide!
</details>

### What browsers are you seeing the problem on?

N/A (all of them)

### Relevant log output

```shell
src/App.tsx:35:5 - error TS2322: Type 'Resolver<{ optional: number | undefined; }>' is not assignable to type 'Resolver<{ optional?: number | undefined; }, any>'.
  Types of parameters 'values' and 'values' are incompatible.
    Type '{ optional?: number | undefined; }' is not assignable to type '{ optional: number | undefined; }'.

35     resolver: yupResolver(schemaOptionalOnly), // <-- error
       ~~~~~~~~

src/App.tsx:44:5 - error TS2322: Type 'Resolver<{ required: number; optional: number | undefined; }>' is not assignable to type 'Resolver<{ optional?: number | undefined; required: number; }, any>'.
  Types of parameters 'values' and 'values' are incompatible.
    Type '{ optional?: number | undefined; required: number; }' is not assignable to type '{ required: number; optional: number | undefined; }'.

44     resolver: yupResolver(schemaBoth), // <-- error
       ~~~~~~~~
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by leapful** [comment, marked as answer]
@milotoor 

You need to specify the type object on `yupResolver<TypeSchema>` function since the type of schema object and type object are different after using `yup.InferType`.

For example:

```
resolver: yupResolver<yup.InferType<typeof schemaOptionalOnly>>(schemaOptionalOnly);
```

**Reply by milotoor** [reply]
@leapful thanks for the response, I'm surprised I didn't think of that.
